visallo/visallo
# Have the client handle session expiration

<img width="500" alt="screen shot 2018-01-02 at 10 16 38 am" src="https://user-images.githubusercontent.com/7098/34490631-0fac99da-efa6-11e7-989b-34ef0cb73b0e.png">

Allows us to timeout the session based on user activity instead of client network activity. New Configuration for setting timeout `web.ui.auth.token.expiration_minutes=60`. Must be >= 1

* Events that extend timeout: (mouse|key)(down|up)
* When a timeout occurs a new overlay (similar to offline) is displayed with a button to continue using Visallo.
* Client monitors actual token expiration using response header, and sends heartbeat requests as needed so authentication token doesn't expire before user activity should. Otherwise clicking "Continue Session" would sometimes be a 403.
* Expiration Date is saved in local storage, so new windows/tabs can extend the date and not expire all windows if any one is idle.


Testing Instructions: 
Ensure logout screen appears when no mouse/key events after expiration minutes. The warning screen will show 30 seconds before. Try with multiple windows/tabs (same session) not different browser, incognito, those will be new sessions that are independent.

Multiple windows will show the warning at about the same time and clicking continue in any one should close the warning in all and extend the time. Also just loading a new tab of the same session will extend all others (just loading is considered some activity).

Offline overlay should override the session overlay. If the system is offline the session warning won't appear and if the warning is showing and the system goes down the offline overlay will replace the session one.

Points of Regression: Login, Offline Overlay, Logout

CHANGELOG
Changed: Session expiration configuration changed to `web.ui.auth.token.expiration_minutes`.
Fixed: More accurate session expiration that uses user activity instead of requests to server. With new session expiration overlay that allows users to extend their session.
